### ISSUE TITLE
Non-reproducible results with `SMS_R_Ld5.ne4_ne4.FSCM-ARM97.pm-cpu_gnu` after newer GNU compilers on `pm-cpu`

### ISSUE THREAD

**Issue by ndkeen** [E3SM-Project/E3SM#6556]
A few weeks ago, we updated the GNU compiler version on pm-cpu https://github.com/E3SM-Project/E3SM/pull/6517. We expected results to be different and we blessed them.  However, one of the tests continues to DIFF.  Trying the test manually, while the atm.log values are identical, when I use `cprnc` on the `eam.h0` netcdf files, they are different between two runs.

```
SMS_R_Ld5.ne4_ne4.FSCM-ARM97.pm-cpu_gnu.eam-scm
SMS_R_Ld5.ne4_ne4.FSCM-ARM97.pm-cpu_gnu
```

It only happens with GNU (Intel still OK). 
And with GNU and DEBUG, I can run 2 cases and see identical netcd files.
I can work toward trying to find if there is a certain file that we can downgrade optimization to still allow BFB between runs.
Note that simple changing `SMS` to `ERS` will not work for this case (for some reason).

I just verified that without the `eam-scm` modifier, I still see a diff in the `eam.h0` file

Example of cprnc command I'm trying: 
```
perlmutter-login11% pwd
/pscratch/sd/n/ndk/e3sm_scratch/pm-cpu/nexty-aug9
perlmutter-login11% cprnc SMS_R_Ld5.ne4_ne4.FSCM-ARM97.pm-cpu_gnu.r00/run/SMS_R_Ld5.ne4_ne4.FSCM-ARM97.pm-cpu_gnu.r00.eam.h0.1997-06-19-84585.nc SMS_R_Ld5.ne4_ne4.FSCM-ARM97.pm-cpu_gnu.r01/run/SMS_R_Ld5.ne4_ne4.FSCM-ARM97.pm-cpu_gnu.r01.eam.h0.1997-06-19-84585.nc | tail -40

                   1   0.129433436191206E-09   0.129433436191206E-09
                   1  (     1,     1) (     1,     1)
          avg abs field values:    1.294334361912064E-10
                                   1.294334361912064E-10
 vap_ice_exchange   (ncol,lev,time)  t_index =     61    61
                  80  (     0,    54,     1) (     0,    58,     1)
                  80   0.625710816493097E-08  -0.118322907027846E-07
                  80   0.625710816493097E-08  -0.118322907027846E-07
                  80  (     0,    54,     1) (     0,    58,     1)
          avg abs field values:    6.217725223578441E-10
                                   6.217725223578441E-10
 vap_liq_exchange   (ncol,lev,time)  t_index =     61    61
                  80  (     1,     1,     1) (     0,    64,     1)
                  80   0.000000000000000E+00  -0.288254087266182E-08
                  80   0.000000000000000E+00  -0.288254087266182E-08
                  80  (     1,     1,     1) (     0,    64,     1)
          avg abs field values:    1.553264739051485E-10
                                   1.553264739051485E-10
 vap_liq_exchange_clubb   (ncol,lev,time)  t_index =     61    61
                  80  (     0,    65,     1) (     0,    60,     1)
                  80   0.393390433956983E-14  -0.140163358697265E-08
                  80   0.393390433956983E-14  -0.140163358697265E-08
                  80  (     0,    65,     1) (     0,    60,     1)
          avg abs field values:    1.752053710446511E-11
                                   1.752053710446511E-11
************************************************************************************************************************************
  
SUMMARY of cprnc:
 A total number of  18134 fields were compared
          of which      1 had non-zero differences
               and      0 had differences in fill patterns
               and      0 had different dimension sizes
               and      0 had different data types
 A total number of    122 fields could not be analyzed
 A total number of      0 time-varying fields on file 1 were not found on file 2.
 A total number of      0 time-constant fields on file 1 were not found on file 2.
 A total number of      0 time-varying fields on file 2 were not found on file 1.
 A total number of      0 time-constant fields on file 2 were not found on file 1.
  diff_test: the two files seem to be DIFFERENT
```

I also just tried a test without `_R` test option, but `_P1` instead and still see same issue. ie:
```
SMS_P1_Ld5.ne4_ne4.FSCM-ARM97.pm-cpu_gnu
```

**Comment by rljacob** [comment]
Here is the relevant output from cprnc:
```
 area   (ncol)
          1        1  (     1) (     1) (     1) (     1)
                   1   2.532510540886779-315   2.532510540886779-315 1.4-316  2.532510540886779-315 5.2E-02  2.532510540886779-315
                   1   2.671960900449453-315   2.671960900449453-315          2.671960900449453-315          2.671960900449453-315
                   1  (     1) (     1)
          avg abs field values:    2.532510540886779-315    rms diff: 0.0E+00   avg rel diff(npos):  5.2E-02
                                   2.671960900449453-315                        avg decimal digits(ndif):  1.3 worst:  1.3
 RMS area                             0.0000E+00            NORMALIZED  0.0000E+00
```
The area is not initialized and is being filled with junk when its output to netcdf.  Different junk each time.

**Comment by rljacob** [comment]
@ndkeen the piece of cprnc output you including in the issue doesn't have enough lines.  Try "tail -40"

**Comment by ndkeen** [comment]
It does make sense that it's a classic case of un-initialized values.

**Comment by bogensch** [comment]
All, I am able to reproduce this problem with gnu compiler and I've linked to the spot where the area in scm is defined.  However, I can't figure out what the heck is going on.  Maybe one of you smarter gentlemen can?

The relevant code:

[https://github.com/E3SM-Project/E3SM/blob/69f24bfac5823a12e5950555bb23500338a23220/components/eam/src/dynamics/se/dyn_grid.F90#L666-L679](url)

using print statements I've verified that the variable "area_scm" is an actual value (something like ~2.e-03).  The value "area_scm" is then passed to the add_cam_grid_attribute_1d_r8 subroutine.  Within this subroutine I also verified that the "values" array still contains an actual value right before line 1953.  I imagine something is going haywire with the 1d array and the interfaced routine.... but not sure what or how to fix it....

[https://github.com/E3SM-Project/E3SM/blob/69f24bfac5823a12e5950555bb23500338a23220/components/eam/src/utils/cam_grid_support.F90#L1932-L1964](url)

**Comment by ndkeen** [comment]
OK that helps a little.

After some testing, I see that with GNU optimized builds, I can reduce opt levels in one file to get BFB behavior.
If we add this to `cime_config/machines/Depends.pm-cpu.gnu.cmake`

```
set(ADJUST
eam/src/dynamics/se/dyn_grid.F90
)

# now with SMS_R_Ld5.ne4_ne4.FSCM-ARM97.pm-cpu_gnu, trying:
if (NOT DEBUG)
  foreach(ITEM IN LISTS ADJUST)
    e3sm_add_flags("${ITEM}" "-O")
  endforeach()
endif()
```

then two consecutive runs of `SMS_D_R_Ld5.ne4_ne4.FSCM-ARM97.pm-cpu_gnu` will have cprnc identical eam.h0 files.

Ideally, we could find what is happening inside the file to cause the issue.

**Comment by ndkeen** [comment]
When I do a ncdump of the relevant netcdf files when they differ, the only difference I see is area:

```
<  area = 3.20013922975731e-315 ;
---
>  area = 3.58791503618982e-315 ;
```

**Comment by ambrad** [comment]
@ndkeen I think you've spotted the issue. `pearea` is a persistent, allocated array that `attr` will point to (see utils/cam_grid_support.F90, routine `cam_grid_attr_init_1d_r8`). In contrast, `area_scm` is a local array.

A possible workaround is simply to use `pearea` and set every entry to `area_scm(1)`, then pass `pearea`, `pemap` as usual to the routine. This will work only if `pearea` gets initialized in the SCM case. It looks like it might from cursory inspection, but I'm far from sure.

**Comment by ndkeen** [comment]
Thanks.

If I'm following, I tried 
```
      area_scm(1) = 1.0_r8 / elem(1)%rspheremp(1,1)
      !ndk      call cam_grid_attribute_register(trim(gridname), trim(areaname),   &                                                                                                                                       
      !ndk                               'gll grid areas', trim(ncolname), area_scm)                                                                                                                                       

      pearea=area_scm(1)
      call cam_grid_attribute_register(trim(gridname), trim(areaname),   &
           'gll grid areas', trim(ncolname), pearea, pemap)
```

but get error:
```
0: PIO: WARNING:  Invalid count array passed to PIO_get_var() function, pio_get_vara_1d_double(). The size of count array (           2  elements) is greater than the number of dimensions (           1 ) of the variable. The excess number of elements in the start/\
count arrays will be ignored. varid =           70 , file id =           99, (spio_get_var.F90:519)
0: PIO: ERROR: Writing variable area, varid=2, (total number of variables = 1) to file SMS_R_Ld5.ne4_ne4.FSCM-ARM97.pm-cpu_gnu.peareaA.eam.h0.1997-06-19-84585.nc (ncid=130) using serial I/O failed.. NetCDF: Start+count exceeds dimension bound (error num=-57), (/gl\
obal/cfs/cdirs/e3sm/ndk/repos/nexty-aug9/externals/scorpio/src/clib/pio_darray_int.c:984)
...
```

**Comment by ambrad** [comment]
For the allocatable approach, you'll need to make it `allocatable, pointer`, not just `allocatable`. That's because the memory needs to be persistent since `attr` will point to it. An alternative is to put it at module scope like `pearea`.

**Comment by ndkeen** [comment]
Yea it wouldnt let me compile with `allocatable,pointer`.  I did just try this and seems to have worked.

```
! at module level:
real(r8),          pointer :: scm_pearea(:) => null() ! SCM pe-local areas
...
       allocate(scm_pearea(1)) !ndk                                                                                                                                                                                  
       scm_pearea(1) = 1.0_r8 / elem(1)%rspheremp(1,1)
       call cam_grid_attribute_register(trim(gridname), trim(areaname), &
            'gll grid areas', trim(ncolname), scm_pearea, pemap)
...
  nullify(scm_pearea) !ndk 
```

**Comment by ndkeen** [comment]
I have a branch `ndk/eam/fix-setting-writing-of-area-for-SCM` that I tested. I think we can use REP tests here (had issues with ERS tests, so had been doing two SMS tests and comparing netcdf files with cprnc).

With this branch, I verified the following pass:
```
REP_R_Ld5.ne4_ne4.FSCM-ARM97.pm-cpu_gnu
REP_R_Ld5.ne4_ne4.FSCM-ARM97.pm-cpu_gnu.eam-scm
REP_D_R_Ld5.ne4_ne4.FSCM-ARM97.pm-cpu_gnu.eam-scm
```

I guess I could make a PR with this and we could iterate. And maybe wait for Peter B to chime in?
